ENH: Implement np.strings.slice as a gufunc

ArvidJB opened the pull request:
This commit adds a `np.strings.slice` function which vectorizes slicing of string arrays. For example
```
    >>> a = np.array(['hello', 'world'])
    >>> np.char.slice(a, 2)
    array(['he', 'wo'], dtype='<U5')
```

This supports fixed-length and variable-length string dtypes. It also supports broadcasting the start, stop, step args:
```
    >>> b = np.array(['hello world', 'γεια σου κόσμε', '你好世界', '👋 🌍️'], dtype=np.dtypes.StringDType())
    >>> np.strings.slice(b, [3, 0, 2, 1], -1)
    array(['lo worl', 'γεια σου κόσμ', '世', ' 🌍'], dtype=StringDType())
```

Closes #8579

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

ArvidJB committed:
ENH: Implement np.strings.slice as a gufunc

This commit adds a `np.strings.slice` function which
vectorizes slicing of string arrays. For example
```
    >>> a = np.array(['hello', 'world'])
    >>> np.char.slice(a, 2)
    array(['he', 'wo'], dtype='<U5')
```

This supports fixed-length and variable-length string dtypes.
It also supports broadcasting the start, stop, step args:
```
    >>> b = np.array(['hello world', 'γεια σου κόσμε', '你好世界', '👋 🌍️'], dtype=np.dtypes.StringDType())
    >>> np.strings.slice(b, [3, 0, 2, 1], -1)
    array(['lo worl', 'γεια σου κόσμ', '世', ' 🌍'], dtype=StringDType())
```

Closes #8579

ArvidJB commented:
Thanks @seberg for taking a look!

I had a couple of questions:
1. is the API design acceptable? I intentionally made `slice` only take positional arguments to mimic `builtins.slice`'s behavior. Or should we support positional arguments, like `np.strings.slice(a, step=-1)`?
2. For variable-length strings we end up doing three passes over the string: (1) to compute `num_codepoints`, (2) to compute `outsize`, (3) to copy each codepoint. I cannot see a way to avoid this, but maybe someone has an idea?
3. Negative steps, where "step < 1" are tricky, because we could potentially read past the beginning of the string by accident. I would be thankful if someone could double-check the logic I wrote there.

seberg commented:
We can always start with only positional and relax it if someone feels strongly (I also wondered briefly if we should accept slices, but probably better not).

I am hoping @lysnikolaou or @ngoldbaum will have time to dive in (but I would not expect this to happen very quickly necessarily).
In general, I would think you could at least bring it down to 2 passes (merge the first two), although it may need a custom new method on the buffer helper implementation?
(I'll assume that small, usually one, steps are typically, and we shouldn't worry about optimizing large steps much.)

ngoldbaum commented:
Neat, thanks for the ping. I'd like to spend some time looking closely at this. Do you mind if I push directly to this branch with fixes?

pypy has a neat optimization that allows O(1) indexing into UTF-8 strings by storing effectively a reverse index into the string. There are also some clever tricks to bring the memory overhead down. Adopting something like that would help for performance on operations like this that assume indexing into strings is cheap.

ArvidJB commented:
> Neat, thanks for the ping. I'd like to spend some time looking closely at this. Do you mind if I push directly to this branch with fixes?

Sure, that sounds fine to me!

> pypy has a neat optimization that allows O(1) indexing into UTF-8 strings by storing effectively a reverse index into the string. There are also some clever tricks to bring the memory overhead down. Adopting something like that would help for performance on operations like this that assume indexing into strings is cheap.

Oh, that's clever! Curious to see the details. We should probably add some benchmarks? It's probably worth to have a dedicated codepath for `step==1`.

ngoldbaum commented:
So there's a repo here: https://github.com/pypy/fast-utf8-methods, the pypy UTF-8 code also lives in pypy here: https://github.com/pypy/pypy/tree/6742499bbf3fc0aa63702fe4aa27147e11050c74/rpython/rlib/fastutf8

@mattip might know if there's a technical explanation somewhere of how this works. I'm not aware of a blog post or paper.

This is more of a long-term idea. I'm not sure if the memory overhead of building the index is worth it - not everyone wants to do fast string indexing by character.

mattip commented:
> I'm not sure if the memory overhead of building the index is worth it - not everyone wants to do fast string indexing by character.

I don't think there is a technical summary of the SIMD utf8 code. The code actually was never merged into PyPy, it was an experiment that one developer tried to push forward but was never completed. PyPy does [build an index](https://github.com/pypy/pypy/blob/6742499bbf3fc0aa63702fe4aa27147e11050c74/rpython/rlib/rutf8.py#L516) of every 4th codepoint which is only allocated and built on-demand when a utf-8 string is sliced.

ngoldbaum commented:
Reminder to myself to look through this sometime next week.

ngoldbaum commented on numpy/_core/strings.py:
```diff
@@ -1638,3 +1639,89 @@ def translate(a, table, deletechars=None):
+    if np.any(step == 0):
+        raise ValueError("slice step cannot be zero")
+
+    if start is None:
+        start = np.where(step < 0, np.iinfo(np.intp).max, 0)
+    start = np.asanyarray(start)
+    if not np.issubdtype(start.dtype, np.integer):
+        raise TypeError(f"unsupported type {start.dtype} for operand 'start'")
+
+    if stop is None:
+        stop = np.where(step < 0, np.iinfo(np.intp).min, np.iinfo(np.intp).max)
+    stop = np.asanyarray(stop)
+    if not np.issubdtype(stop.dtype, np.integer):
+        raise TypeError(f"unsupported type {stop.dtype} for operand 'stop'")
```
```suggestion
    if start is None:
        start = np.where(step < 0, np.iinfo(np.intp).max, 0)
    else:
        start = np.asanyarray(start)
        if not np.issubdtype(start.dtype, np.integer):
            raise TypeError(f"unsupported type {start.dtype} for operand 'start'")

    if stop is None:
        stop = np.where(step < 0, np.iinfo(np.intp).min, np.iinfo(np.intp).max)
    else:
        stop = np.asanyarray(stop)
        if not np.issubdtype(stop.dtype, np.integer):
            raise TypeError(f"unsupported type {stop.dtype} for operand 'stop'")
```

no need to call `asanyarray` or check dtypes if you create an array from `np.where`, I think.

ngoldbaum reviewed (commented):
I didn't go through the C++ code yet but I spotted two issues in the python layer on my first pass

ngoldbaum commented on numpy/_core/src/umath/stringdtype_ufuncs.cpp:
```diff
@@ -2142,6 +2142,238 @@ string_inputs_promoter(
+            }
+
+            PyMem_RawFree(buf);
+        }
```
we don't allow in-place operations yet so this bit is unnecessary

ngoldbaum commented on numpy/_core/src/umath/stringdtype_ufuncs.cpp:
```diff
@@ -2142,6 +2142,238 @@ string_inputs_promoter(
+            inbuf.buffer_memcpy(outbuf, outsize);
+        }
+        else {
+            // general case, iterate over each characterof the string
```
```suggestion
            // general case, find each character of the output string and copy it into the output
```

ngoldbaum commented on numpy/_core/src/umath/stringdtype_ufuncs.cpp:
```diff
@@ -2142,6 +2142,238 @@ string_inputs_promoter(
+                              "Failed to allocate string in slice");
+                goto fail;
+            }
+        }
```
we're not allowing in-place operations yet so you can just delete the case above. In principle if we add it in the future we'd need to redo this but better to do that than have dead code IMO.

I know there are other places we have dead code you're copying from, but was probably a mistake on my part to allow that, better to not add new dead code.

ngoldbaum commented on numpy/_core/src/umath/stringdtype_ufuncs.cpp:
```diff
@@ -2142,6 +2142,238 @@ string_inputs_promoter(
+        // move to start position in inbuf
+        inbuf += start;
+
+        // first pass: compute outsize
```
On the first pass over the string, if `step > 0`, couldn't you save the byte indices in the input string for each character in the output string, avoiding the need to do a second scan over the whole string?

Then, once you know outsize, you allocate the output buffer and copy each string out using the index array you find earlier. You could also save the size of each character in bytes but that's not necessary - it turns out that you only need to look at the first byte of a UTF-8 character to see how many bytes the character is.

ArvidJB replied:
I think we can probably use a thread-local buffer (or `std::vector`) here which can avoid the need to allocate on every run. I'll try to implement that.

ArvidJB replied:
I implemented your suggestion. I did not use a thread-local buffer since I the buffer will already be reused across all elements in the array.

ngoldbaum reviewed (commented):
I think it is necessary to do this in two passes for UTF-8 but I think you only need to do one pass over the whole string if you allocate another temporary buffer to build and save a byte index from the input string into the output string on the first pass.

I don't think it makes sense to build and cache an index for the *whole* string, but there's no need to calculate the index of each character in the output string twice.

ngoldbaum commented:
This also needs a release note.

ArvidJB commented on numpy/_core/defchararray.py:
```diff
@@ -1209,6 +1211,19 @@ def isdecimal(self):
         """
         return isdecimal(self)
 
+    def slice(self, start=None, stop=None, step=None, /):
```
I assume that means I would import `slice` directly in `numpy/char/__init__.py`?

ArvidJB replied:
(I have removed `slice` from `CharArray` already)

ngoldbaum replied:
On second thought the way you have it in this PR now is better. We shouldn't add new functions to `np.char` if we're going to deprecate it sometime in the next year or two and any new code can just use the `np.strings` path.

ArvidJB replied:
Okay, I will remove it from `numpy.char` altogether.

ngoldbaum replied:
Awesome, please wait for me to push here. I'm going to push to your branch shortly.

ngoldbaum commented:
Just a head's up I'm on Christmas break right now. I'm
planning to look at this in the new year.

ngoldbaum commented on numpy/_core/src/umath/stringdtype_ufuncs.cpp:
```diff
@@ -2142,6 +2144,171 @@ string_inputs_promoter(
+                num_codepoints++;
+                int num_bytes = num_bytes_for_utf8_character(
+                        (const unsigned char *)inbuf_ptr);
+                codepoint_offsets.push_back(inbuf_ptr - is.buf);
```
It occurs to me that you only need to store this if it isn't one byte. Since we assume ASCII/Latin-1 is the common case, we can probably get some performance back there by only storing offsets for non-Latin-1 bytes.

I think I can implement this today and just push to your branch.

I like using a `std::vector` here, and we can probably reduce the heap thrashing a lot for common scientific data just by implementing my above suggestion. I agree doing things with a thread-local here doesn't make sense.

ngoldbaum replied:
I think we can also compute `outsize` in this loop...

ngoldbaum replied:
I spent a little while on this and came up with the following design. Instead of using a vector of ints, you can use a vector of pairs of ints, where each pair is a codepoint index and offset into the array. Then you can compute outsize by first assuming that the size of the string is the number of codepoints in the slice (e.g. one byte per character) and then go through the offsets vec and for each multibyte character that is in the slice, add the extra bytes for that character to `outsize`. That avoids a pass over the whole string.

ngoldbaum replied:
On further reflection I don't think this is actually better than what you have here. Your approach is a lot simpler. I've refactored things a bit though and done some cleanups. I'm hoping to push something to your branch shortly.

ArvidJB replied:
Thanks for looking into this! As you had alluded to in your first comment, do you want to take a stab at implementing that idea?

Regarding heap-thrashing: I was worried about this at first, but since the `std::vector` is reused across the loop over all array elements I think it is unlikely to be resized many times. For that you would need an array where every string is larger than all previous strings, I think.

ngoldbaum replied:
This ended up reducing down to a much smaller cleanup after I decided this approach was too complicated. I've marked this as resolved but please look over my last push.

ArvidJB replied:
Looks good to me! I agree, your change simplifies the code.

I think it makes sense to keep this "simpler" implementation for now and worry about further optimizing performance only if really necessary. It's already massively faster for my use case:
```
>>> a = np.full(1_000_000, "😀😃😄😁😆😅🤣😂🙂🙃", dtype='T')
>>> start = time.time(); _ = np.array([s[3:7] for s in a], dtype=a.dtype); end = time.time()
>>> end - start
0.4303429126739502
>>> start = time.time(); _ = np.strings.slice(a, 3, 7); end = time.time()
>>> end - start
0.08485579490661621
```

ngoldbaum commented on numpy/char/__init__.py:
```diff
@@ -1,2 +1,5 @@
 from numpy._core.defchararray import *
+# import slice function directly from strings module instead
+# of adding it to the deprecated defchararray module
+from numpy._core.strings import slice
```
Oh oops missed this. On second thought I think we should just delete it. If you really do feel strongly that having the function be available in the `np.char` namespace is important, you should add it to `__all__` too.

ngoldbaum replied:
It's not really that `defchararray.py` is deprecated, we want to get rid of the entire `np.char` namespace since it's also associated with the CharArray concept which we want to move away from.

ngoldbaum commented:
@lysnikolaou is there any chance I can get you to do a high-level once-over on this PR? There's no need to go through the new C++ code in detail, I just did that.

ngoldbaum commented:
@mhvk could you give this one a once-over if you have a little time?

mhvk commented on numpy/_core/src/umath/stringdtype_ufuncs.cpp:
```diff
@@ -2142,6 +2144,156 @@ string_inputs_promoter(
+        // compute outsize
+        npy_intp outsize = 0;
+        for (int i = start; step > 0 ? i < stop : i > stop; i += step) {
+            outsize += num_bytes_for_utf8_character(codepoint_offsets[i]);
```
This seems weird - you already calculated the number of bytes - can one not just take a difference between the relevant offsets? (Alternatively, perhaps there should also be a `std::vector::<unsigned char> code_point_lengths`?)

ngoldbaum replied:
I didn't want to add the extra allocation because this function is so simple it seemed better to just call it twice. That said the allocation is amortized over the array so perhaps it doesn't matter.

What about making `num_bytes_for_utf8_character` a `static inline` function? It's can also make it branchless, I think, with some bitmath...

mhvk replied:
Yes, you're right, it should be possible to count the number of bytes in UTF8 by bitmath; after all, all information on a charcter's length is in the first byte (e.g., http://www.daemonology.net/blog/2008-06-05-faster-utf8-strlen.html)

mhvk replied:
p.s. Totally fine to leave this for a subsequent PR!

ArvidJB replied:
There's also https://github.com/simdutf/simdutf. But I will leave all of this as an exercise for the reader....

ArvidJB replied:
Here's a link to the codepoint counting implementation:
https://github.com/simdutf/simdutf/blob/c4c8e0c09c8b65d4d729ae7192f62ae1eac24b4d/src/generic/utf8.h#L10

ArvidJB replied:
And for AVX512: https://github.com/simdutf/simdutf/blob/c4c8e0c09c8b65d4d729ae7192f62ae1eac24b4d/src/icelake/implementation.cpp#L1272

mhvk commented on numpy/_core/src/umath/stringdtype_ufuncs.cpp:
```diff
@@ -2142,6 +2144,156 @@ string_inputs_promoter(
+        buf = (char *)os.buf;
+
+        if (outsize > 0) {
+            if (step == 1) {
```
This case could be taken out earlier, since in that most common case calculating `outsize` is trivial too.

mhvk commented on numpy/_core/src/umath/stringdtype_ufuncs.cpp:
```diff
@@ -2142,6 +2144,156 @@ string_inputs_promoter(
+                memcpy(buf, codepoint_offsets[start], outsize);
+            }
+            else {
+                npy_intp i_idx = start, o_idx = 0;
```
Nitpick, but why not just
```
for (npy_intp o_idx = 0, i_idx = start; o_idx < slice_length; o_idx++, i_idx += step) {
```

mhvk commented on numpy/_core/code_generators/ufunc_docstrings.py:
```diff
@@ -5314,3 +5314,58 @@ def add_newdoc(place, name, doc):
+    ----------
+    a : array-like, with ``StringDType``, ``bytes_``, or ``str_`` dtype
+
+    start : the start of the slice, can be None, an integer or
```
For the `ufunc`, it is not true that `start`, `stop` and `step` can be None. Please adjust.

I think I'd just refer to docstring for `np.strings.slice` for examples.

mhvk commented on numpy/_core/strings.py:
```diff
@@ -1638,3 +1639,91 @@ def translate(a, table, deletechars=None):
+        if not np.issubdtype(stop.dtype, np.integer):
+            raise TypeError(f"unsupported type {stop.dtype} for operand 'stop'")
+
+    a = np.asanyarray(a)
```
This is not necessary - the ufunc machinery will do this already.

ArvidJB replied:
Removing this `asanyarray` works.

mhvk commented on numpy/_core/strings.py:
```diff
@@ -1638,3 +1639,91 @@ def translate(a, table, deletechars=None):
+    # adjust start, stop, step to be integers, see logic in PySlice_Unpack
+    if step is None:
+        step = 1
+    step = np.asanyarray(step)
```
Here and below - these `asanyarray` are not really necessary - the ufunc will do it for you. The only downside would be that you'd get a loop-not-found error if people pass in non-integer inputs.

ArvidJB replied:
I am probably missing something, but if I remove those `asanyarray` calls I get errors like
```
FAILED numpy/_core/tests/test_strings.py::TestMethodsWithUnicode::test_slice[args17-T] - numpy._core._exceptions._UFuncNoLoopError: ufunc '_slice' did not contain a loop with signature matching types (<class 'numpy.dtypes.StringDType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes.Int64DType'>, <class 'numpy.dtypes._PyLongDType'>) -> None
```

ngoldbaum replied:
I think we might need promoters to PyInt. We needed that elsewhere...

seberg replied:
Maybe an OK work-around.  The problem is more broad though: without promotion, e.g. `int32` input will also fail, we just mostly care about scalars, so you don't notice.
I.e. you really would need `asarray(step).astype(np.intp, copy=False, casting="same_kind")` to support small integer sizes as well, unfortunately.

You could probably even register the loop with `PyArray_IntAbstractDType` to begin with, but I am not sure we are letting you do that yet.
Otherwise, need to add a promoter with it right now, there should be examples.

(Yes, dealing with different integer sizes is tedious, especially since with `int64` and `uint64` you have two "maximal" types...)

seberg replied:
I would agree that we should possibly have a standard promoter one can easily enable that just converts python integer to default integer.
(That is, we can always do that I think, except when we want to use `result_type` style promotion, but that only makes sense for type homogenous signatures.)

mhvk replied:
It sounds like it may be a bit complicated to add the promotor -- would be nice to have a standard for that! -- so for now perhaps we could simply do,
```
np.asanyarray(..., dtype=np.intp)
```
and add a test that it works with, say, `int8` and `uint64`. I think the above only fails for `uint64` above `2**63`, and maybe we should simply not worry about that?

ArvidJB replied:
I'm not sure there is a "correct" way to handle `uint64`:
```
In [5]: a = np.array([123, 2**64 - 1], dtype='uint64')

In [6]: a
Out[6]: array([                 123, 18446744073709551615], dtype=uint64)

In [7]: np.asanyarray(a, dtype=np.intp)
Out[7]: array([123,  -1])

In [8]: a.astype(np.intp, casting='same_kind')
Out[8]: array([123,  -1])

In [9]: a.astype(np.intp, casting='safe')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[9], line 1
----> 1 a.astype(np.intp, casting='safe')

TypeError: Cannot cast array data from dtype('uint64') to dtype('int64') according to the rule 'safe'

In [11]: np.array([123, 456], dtype='uint64').astype(np.intp, casting='safe')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[11], line 1
----> 1 np.array([123, 456], dtype='uint64').astype(np.intp, casting='safe')

TypeError: Cannot cast array data from dtype('uint64') to dtype('int64') according to the rule 'safe'
```
Ideally there would be a, say, `casting='check'` option which checks if elements are `<= 2**63 - 1`, but it does not seem to exist?

There's a related discussion on https://github.com/numpy/numpy/issues/8987 which I found, but that is asking for a way to check all operations as well, which is much more than we would need here.

seberg replied:
Yes, uint64 is impossible to get 100% (but I am not sure it matters in practice, since you need absurd numbers).

I don't love the verbosity/boilerplate of the promoter, but since there are already a few, I hope it may not be annoying to implement.
But, I don't care too much, it would be nice if the promoter really always picks the one loop, then the loop might as well just be selected for all integers to begin with (and I guess that doesn't work yet, but it should).

A cast which checks could indeed be a neat thing, but is a whole bigger thing that needs some thinking (implementing isn't so hard I think, more about how to thread it into the ufunc machinery).

seberg replied:
Didn't have a look, but tests are failing on 32bit machines.  I think the reason is probably just that you used `int64` in the promoter.
That might actually be better, but in that case, you need to change `intp` to `int64` in all loops.

(Yes, `intp` is sufficient and great and even the default, so either version has downsides without a "safe cast".)

ArvidJB replied:
I did not even know that 32 bit computers still existed! Should be fixed now...

mhvk commented on numpy/_core/tests/test_strings.py:
```diff
@@ -859,6 +859,37 @@ def test_rpartition(self, buf, sep, res1, res2, res3, dt):
+                        for s, arg in zip(buf, zip(*bcast_args))],
+                       dtype=dt)
+        assert_array_equal(act, res)
+
```
Would be good to test failures too (i.e., wrong input).

ArvidJB replied:
I added two simple wrong input examples. Let me know if I should add more.

mhvk reviewed (commented):
This looks very nice! I did still have a few comments, see in-line.

mhvk reviewed (approved):
Two utter nitpicks, and one query as I'm confused why the `asanyarray` would be needed - but clearly I'm missing something too...

None of this is important, so I'll approve now (though probably need to squash-merge given the large number of commits).

ArvidJB commented:
I rebased on main and squashed the commits. Is this ready to be merged?

seberg commented:
> I rebased on main and squashed the commits. Is this ready to be merged?

I'll have Nathan have another look over, but likely.  The CircleCI failure is real and we should address it (a tiny thing): The documentation is slightly wrong somewhere:
```
1669     Examples
1670     --------
1671     >>> import numpy as np
1672     >>> a = np.array(['hello', 'world'])
1673     >>> np.char.slice(a, 2)
UNEXPECTED EXCEPTION: AttributeError("module 'numpy.char' has no attribute 'slice'")
```

ArvidJB committed:
Fix reference to numpy.char.slice

ArvidJB commented:
> > I rebased on main and squashed the commits. Is this ready to be merged?
> 
> I'll have Nathan have another look over, but likely. The CircleCI failure is real and we should address it (a tiny thing): The documentation is slightly wrong somewhere:
> 
> ```
> 1669     Examples
> 1670     --------
> 1671     >>> import numpy as np
> 1672     >>> a = np.array(['hello', 'world'])
> 1673     >>> np.char.slice(a, 2)
> UNEXPECTED EXCEPTION: AttributeError("module 'numpy.char' has no attribute 'slice'")
> ```

Ah, looks like doctest running was fixed in one of the commits included in the rebase!

We had decided to not add `np.char.slice` since it's supposedly a deprecated module, but I had not fixed the doctests.

Everything is passing now.

ngoldbaum commented:
Thanks for your patience on getting this reviewed and merged @ArvidJB. If you spot other bugs or missing features in `np.strings` please don’t hesitate to follow up with more issues or PRs!

mhvk commented:
Thanks, @ArvidJB, really nice!
